site/source/docs/porting/asyncify.rst

```diff
@@ -4,123 +4,109 @@
-      console.log('Hello');
-      setTimeout(function() {
-        console.log('World ' + i + '!');
-        async_return_value = 0;
-      }, 1000);
-    }
+    emcc -O3 example.cpp -s ASYNCIFY
 
-Specifically, a number of aspects should be taken into consideration:
- - Split the function when an async function is called, and the second function should be registered as the callback for the async function
- - Any function that calls an async function also becomes an async function.
- - Keep all local variables available to the callback
- - Closure cannot be used in order to make asm.js validated code.
- - Take care of loops and branches
- - Make the return value available to the callee
- - Some functions could be both sync or async, depending on the input.
+.. note:: It's very important to optimize (``-O3`` here) when using Asyncify, as unoptimized builds are very large.
 
-And the ``ASYNCIFY`` option does all above automatically, through a number of transformations on LLVM IR.
+And you can run it with
 
+::
 
-Usage
-=====
+    nodejs a.out.js
 
-Call ``emscripten_sleep()`` whenever you need to pause the program, and add ``-s ASYNCIFY=1`` to emscripten.
+You should then see something like this:
 
-Sometimes it's a good replacement of ``emscripten_set_main_loop``, you may replace all ``sleep``-alike functions with ``emscripten_sleep``, instead of refactoring the whole main loop.
+::
 
-Also ``emscripten_sleep(1)`` can be used to 'interrupt' your code, such that the JavaScript engine can do the rendering and process events.
+    sleeping...
+    sleeping...
+    sleeping...
+    sleeping...
+    sleeping...
+    timer happened!
 
-Extensions
-==========
+The code is written with an "infinite loop" that sleeps, which normally would not allow async events to be handled by the browser. With Asyncify, those sleeps actually exit to the browser's main event loop, and the timer can happen!
 
-It is possible to implement more new async functions that appears to be sync in C.
+Calling async Web APIs as if they were synchronous
+==================================================
 
-- Implement the function normally in a JavaScript library, suppose the function name is ``func``.
-- Add ``func`` into ``ASYNCIFY_FUNCTIONS``
-- When ``func`` is called and finished, the program will NOT continue, instead it just save the context and exit.
-- Call ``_emscripten_async_resume`` when you want to resume the program, usually in the callback functions of some async calls.
+Aside from ``emscripten_sleep`` and the other standard sync APIs Asyncify supports, you can also add your own functions. To do so, you must create a JS function that is called from wasm (since Emscripten controls pausing and resuming the wasm from the JS runtime). One way to do that is with a JS library function; another is to use ``EM_JS``, which we'll use in this next example:
 
-Please read ``src/library_async.js`` for details.
+::
 
-Limitations
-===========
+    // example.c
```

Comment:
to me this example seems like sort of a special case of the sleep example (just a sleep with a different wakeup condition). Could we use this to call a promise-returning API such as fetch as if it were synchronous?

Reply:
What's tricky with fetch is that it requires a file, so it's not as easy to set up a full example for the user to run.

I can't think of a better promise-based API, and can't see how to search for one. Maybe someone has a good idea?

Reply:
Well I guess it can fetch the html page itself, since that exists. Somewhat artificial but at least it uses a promise. Pushed a commit with that now.

Reply:
@kripken Maybe `crypto.subtle.digest`?